src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_state.go

```diff
@@ -130,3 +118,31 @@ func (st *StoragePowerActorState_I) _sampleMinersToSurprise(rt Runtime, challeng
 
 	return sampledMiners
 }
+
+func (st *StoragePowerActorState_I) _safeGetPowerEntry(rt Runtime, minerID addr.Address) PowerTableEntry {
+	powerEntry, found := st.PowerTable()[minerID]
+
+	if !found {
+		rt.Abort("sm._safeGetPowerEntry: miner not found in power table.")
+	}
+
+	return powerEntry
+}
+
+func (st *StoragePowerActorState_I) _ensurePledgeCollateralSatisfied(rt Runtime) bool {
+
+	// TODO: convert msgSender to MinerActorID
+	var minerID addr.Address
+
+	powerEntry := st._safeGetPowerEntry(rt, minerID)
+	pledgeCollateralRequired := st._getPledgeCollateralReq(rt, powerEntry.ActivePower()+powerEntry.InactivePower())
+
+	if pledgeCollateralRequired < powerEntry.LockedPledgeCollateral() {
+		return true
+	} else if pledgeCollateralRequired < (powerEntry.LockedPledgeCollateral() + powerEntry.AvailableBalance()) {
+		st._lockPledgeCollateral(rt, minerID, (pledgeCollateralRequired - powerEntry.LockedPledgeCollateral()))
+		return true
+	}
```

Comment:
Blocking: this whole construction assumes that pledge collateral requirements only increase, which isn't true (miners can reduce stake, which must free up collateral). It would be much clearer to:
- compute requirement
- compute delta between requirement and locked
- if delta < available, update locked by delta (delta may be negative) and return true

Reply:
will move to what Joe suggested below and only keep one number as the balance if we are in sync.

Reply:
ack